
# HG changeset patch
# User Magnus Melin <mkmelin+mozilla@iki.fi>
# Date 1756331278 14400
# Node ID 1b8b56fae40702a835822aee138c9b18c867f519
# Parent  c12ca715def69fbab90f10e4037fc2176fa9582c
Bug 1982353 - Fix decoding q-encoded empty header. r=darktrojan,freaktechnik,kaie


jsmime handles this correctly (I added tests), but libmime did not.

Differential Revision: https://phabricator.services.mozilla.com/D261092

diff --git a/mailnews/mime/jsmime/test/unit/test_header.js b/mailnews/mime/jsmime/test/unit/test_header.js
--- a/mailnews/mime/jsmime/test/unit/test_header.js
+++ b/mailnews/mime/jsmime/test/unit/test_header.js
@@ -1032,17 +1032,20 @@ define(function (require) {
         // Some basic sanity tests for the test process
         ["Test", "Test"],
         ["Test 2", "Test 2"],
         ["Basic  words", "Basic  words"],
         ["Not a =? word", "Not a =? word"],
 
         // Simple 2047 decodings
         ["=?UTF-8?Q?Encoded?=", "Encoded"],
+        ["=?utf-8?Q??=", ""],
+        ["=?utf-8?Q??=  ", ""], // FWS (folding white space) is semantically insignificant.
         ["=?UTF-8?q?Encoded?=", "Encoded"],
+        [" =?UTF-8?q?EncodedFWS?= ", "EncodedFWS"], // FWS is semantically insignificant.
         ["=?ISO-8859-1?Q?oxyg=e8ne?=", "oxyg\u00e8ne"],
         ["=?UTF-8?B?QmFzZTY0?=", "Base64"],
         ["=?UTF-8?b?QmFzZTY0?=", "Base64"],
         ["=?UTF-8?Q?A_space?=", "A space"],
         ["=?UTF-8?Q?A space?=", "A space"],
         ["A =?UTF-8?Q?B?= C", "A B C"],
         ["=?UTF-8?Q?A?= =?UTF-8?Q?B?=", "AB"],
         ["=?UTF-8?Q?oxyg=c3=a8ne?=", "oxyg\u00e8ne"],
diff --git a/mailnews/mime/src/mimehdrs.cpp b/mailnews/mime/src/mimehdrs.cpp
--- a/mailnews/mime/src/mimehdrs.cpp
+++ b/mailnews/mime/src/mimehdrs.cpp
@@ -41,23 +41,20 @@ void MimeHeaders_convert_header_value(Mi
         opt->default_charset ? nsDependentCString(opt->default_charset)
                              : EmptyCString(),
         output);
     value.Assign(output);
     return;
   }
 
   if (opt && opt->rfc1522_conversion_p) {
-    nsAutoCString temporary;
+    nsAutoCString decoded;
     MIME_DecodeMimeHeader(value.get(), opt->default_charset,
-                          opt->override_charset, true, temporary);
-
-    if (!temporary.IsEmpty()) {
-      value = temporary;
-    }
+                          opt->override_charset, true, decoded);
+    value = decoded;
   } else {
     // This behavior, though highly unusual, was carefully preserved
     // from the previous implementation.  It may be that this is dead
     // code, in which case opt->rfc1522_conversion_p is no longer
     // needed.
     value.Truncate();
   }
 }

